lab/13-npm/classes.js

```diff
@@ -0,0 +1,192 @@
+                    
+                    //if the product is not expired, a pair (object, number)
+                    //that stands for (product, quantity) is pushed into the cart
+                    var pair = [];
+                    pair.push(available_products[i])
+                    pair.push(quantity);
+                    this.shoppingCart.push(pair);
+                    //the for loop is ended - there's no need to keep searching if the product
+                    //has been found
+                    break;
+                }
+        }
+
+        //if the product is never found in the so-called database an error is thrown
+        if (!available)
+            throw 'Product out of stock!';
+        }
+
+    removeFromCart(product){
+        var inCart = false;
+        //parsing the array from the end to the beginning - doing this due to the behavior
+        //of the splice() function. splice() completely removes elements from the array, so
+        //the current index changes. One way of making sure nothing goes wrong (index overflow etc)
+        //is starting at the end
+        //splice() is used instead of delete() because the latter only deletes the element of the array
+        //but does not reindex it, nor does it update its length. the deleted elements appear as undefined,
+        //whereas splice() completele removes the element.
+        for (let i = this.shoppingCart.length - 1; i >= 0; i--){
+                if (product === this.shoppingCart[i][0].name){
+                    inCart = true;
+                    this.shoppingCart.splice(i, 1);
+                }
+        }
+        if (!inCart)
+            throw 'Product not in cart!';
+    }
+
+}
+
+
+class Cashier extends Person {
+    constructor(cnp, totalSales){
+        super(cnp);
+        this.totalSales = totalSales;
+    }
+
+    scanAndCash(customer, category){
+        //the method takes two arguments - a customer and a so-called database
+        //where one cand find the products that are about to be bought.
+        var sum = 0;
+        //searching for a (product, quantity) pair in the shopping cart until
+        //a product name matches a product name in the so-called database
+        //entry[1] means quantity and its value is multiplied by the product price
+        for (const entry of customer.shoppingCart){
+            for (const product of category){
+                if (entry[0].name === product.name){
+                    sum += entry[1] * product.price;
+                }
+            }
+        }
+        //obviously if the customer has less money they can not make the transaction
+        if (sum > customer.money)
+            throw new Error('Insufficient funds!');
+        
+        //the sum is substracted from the wallet, the shopping cart is emptied
+        //and the totalSales is updated to keep track of the income 
+        //and avoid money laundering
+        customer.money -= sum;
```

Comment:
`shoppingCart` and `money` properties should be private and you should only expose them via getters as read-only properties, otherwise any other entity can play with `Customer` internals.

If you want to manipulate them you should implement methods in the `Customer` class and call them where needed.